### ISSUE TITLE
A plain TrainStep loop grows the process by the whole parameter set every step, and dies

### ISSUE THREAD

**Issue by Eric-Louchez** [Shorokoo/Shorokoo#321]
`rig.TrainStep` returns a new `TrainingCheckpoint` whose tensors are `OnnxTensorData<T>` — backend
values whose `Dispose(bool)` body is empty, so the native buffer is released only when the finalizer
runs. The managed wrapper is a handle of a few dozen bytes, so a `TrainStep` loop creates almost no
GC pressure: gen2 never runs on its own, the finalizers never run, and the process grows by
parameters + both optimizer moments **on every step** until it dies.

This is the loop `training.md` documents, and the documentation says nothing about needing help from
the GC to run it.

## Reproduction

Reproduced on `main` (798953d) with the LinuxCPU backend — this is not a CUDA or a Windows artifact.
One `[1024, 1024]` trainable parameter: 4 MiB of parameter, 12 MiB of per-step checkpoint state.

```csharp
[TrainableParamInitializer]
public static partial class Normal02
{
    public static Tensor<float32> Inline(Vector<int64> shape)
        => RandomNormal(shape, Scalar(0f), Scalar(0.02f));
}

[Module]
public partial class BigParamModel
{
    public static Tensor<float32> Inline(Tensor<float32> x)
        => x.MatMul(Normal02.Init([Scalar(1024L), Scalar(1024L)]));
}

var rig = TrainingRig.FromScratch(
    BigParamModel.ComputationGraph, L2Loss.ComputationGraph, AdamWOptimizer.ComputationGraph,
    [new TensorDataModelParam("input", ModelParamType.InputParam, TensorData([4L, 1024L], new float[4096]))],
    new AdamWOptimizerHyperparameters { LearningRate = 0.01f });

var ckpt = rig.CreateInitialCheckpoint();
var proc = Process.GetCurrentProcess();
for (int i = 0; i < 40; i++)
{
    ckpt = rig.TrainStep(ckpt, inputs, targets);
    proc.Refresh();
    Console.WriteLine($"[step {i}] ws {proc.WorkingSet64 >> 20} MiB  " +
                      $"gcheap {GC.GetTotalMemory(false) >> 20} MiB  gen2 {GC.CollectionCount(2)}");
}
```

Observed:

```
[step  0] ws 342 MiB  gcheap 119 MiB  gen2 4
[step  5] ws 413 MiB  gcheap 119 MiB  gen2 4
[step 10] ws 477 MiB  gcheap 119 MiB  gen2 4
[step 15] ws 537 MiB  gcheap 119 MiB  gen2 4
[step 20] ws 605 MiB  gcheap 119 MiB  gen2 4
[step 25] ws 665 MiB  gcheap 119 MiB  gen2 4
[step 30] ws 725 MiB  gcheap 120 MiB  gen2 4
[step 35] ws 785 MiB  gcheap 120 MiB  gen2 4
param tensor type: Shorokoo.OnnxTensorData`1[Shorokoo.float32]
```

~12 MiB/step — exactly parameter + Adam `m` + Adam `v`. The managed heap is flat and **not one gen2
collection happened in 40 steps**, which is why nothing releases anything.

## At model scale

The same shape on a 49,214,208-parameter decoder-only transformer (6 layers, width 384, 6 heads of
64, sequence length 1024, vocabulary 50257) trained through `TrainingRig` with `AdamWOptimizer` on
the CUDA backend: **~565 MiB/step**, matching `49,214,208 × 4 bytes × 3 tensors = 590 MB`. The
process died at step 11 of a planned 48,000 under a 28 GiB cap, and the first symptom was not a
clean out-of-memory but

```
Microsoft.ML.OnnxRuntime.OnnxRuntimeException: [ErrorCode:Fail] bad allocation
   at Shorokoo.Runtime.CompiledGraph.Run(NamedModelParam[] inputs)
   at Shorokoo.TrainingRig.RunStep(...)
   at Shorokoo.TrainingRig.TrainStep(...)
```

with nothing indicating that *host* memory was the issue. On an uncapped 31.8 GiB machine the same
run reaches the physical limit at about step 50.

The workaround is a forced collection after every step:

```csharp
ckpt = rig.TrainStep(ckpt, inputs, targets);
GC.Collect(2, GCCollectionMode.Forced, blocking: true);
GC.WaitForPendingFinalizers();
```

With it, the working set is flat at ~4.2 GiB and step time is unchanged (0.23–0.25 s either way);
5,780 steps ran with steady commit. So the cost of collecting is nil and the loop is sustainable —
but the user has to know to write it.

## Why nothing catches this today

`TrainingMemoryStabilityTests` cannot see it by construction: it gates the **live managed heap**
measured with a forced full collection on both ends, on a `[4, 2]` model, and treats RSS only as a
512 MiB catastrophe backstop. This fault's signature is a flat managed heap, zero gen2 collections
and native growth — every axis the gate deliberately discounts.

Related: #180 (TensorData's ownership model) and #178 (deterministic disposal of ORT handles) cover
the general ownership question. This issue is the specific, fatal, user-facing case on the training
hot path: the documented training loop cannot complete a real run. `FastProcessorHelper.RehostOffSession`
already takes the copy-on-read trade at the two construction-time sites; nothing equivalent happens
per step.

**No migration work.** Shorokoo has no installed base to protect. Change the ownership model, the
checkpoint's tensor representation or the `TrainStep` signature outright — no compatibility shim, no
version bump, no format-version negotiation, and no migration path for checkpoints written by an
earlier build.
